numpy/core/records.py

```diff
@@ -423,6 +423,12 @@ def __new__(subtype, shape, dtype=None, buf=None, offset=0, strides=None,
                                       strides=strides, order=order)
         return self
 
+    def __array_finalize__(self, obj):
+        if type(obj) is not type(self):
+            # invoke __setattr__, which for void dtypes will do
+            # self.dtype = sb.dtype((record, self.dtype))
+            self.dtype = self.dtype
```

Comment:
Maybe I'm being too clever, but I think we want the dtype to be set to `np.record` in the case of an explicit constructor. I still think the clause I have is the right one.

    class TestClass(np.ndarray):
        def __array_finalize__(self, obj):
            print(obj is not None and type(self) is not type(obj), 
                  type(obj) is not type(self))

    a = np.arange(10)

    t1 = a.view(TestClass)  # view fron ndarray
    t2 = TestClass(4)       # explicit constructor
    t3 = t1[:2]             # slice
    t4 = t1.view(TestClass) # view from same class


This prints out (left is your if, right is my if)

    (True, True)
    (False, True)
    (False, False)
    (False, False)

I think the right is what I want. Is that correct?